### PR TITLE
feat: add setting for provider/model-specific parameters

### LINKED ISSUE #6625: [Feature] Add setting for provider/model-specific parameters
### Problem / Use Case

When I use a model, the provider does not gurantee that all parameters in the request payload are accepted/supported. 
A great example of this would be NVIDIA NIM's DeepSeek V4 Flash.
Using AI coding tools with this model and thinking on will result in a
`[400]: Validation: Unsupported parameter(s): thinking`
without a way to strip/remove the parameter, making certain provider/model combinations unusable.

### Proposed Solution

Add a configuration layer to OmniRoute that allows users to define provider-specific or model-specific parameter overrides, including:

- A blocklist/denylist of parameters to strip from outgoing requests (per provider or per model).
- Optionally, an allowlist of parameters to include (defaulting to all).
- A UI or config-based interface to manage these overrides (e.g., integrate.api.nvidia.com: { block_params: ["thinking"] }).

This would allow users to route requests to any provider without worrying about unsupported parameters silently or loudly breaking the request.

As an additional feature, when it receives an `Unsupported parameter` error from upstream provider of the model, OmniRoute should auto-add the parameter to the blocklist/denylist of the provider/model, then retry to prevent the same parameter error in the future.

### Alternatives Considered

- Manually modifying the request payload before sending - fragile, not scalable, and not user-friendly.
- Relying on provider-side error handling - inconsistent and wastes a request round-trip.
- Forking OmniRoute to hardcode exclusions - not maintainable.

### Acceptance Criteria

- A new configuration section allows users to define blocked/allowed parameters per provider or model.
- Requests are automatically filtered before being sent to the upstream provider.
- Invalid parameter errors from providers are suppressed or logged gracefully instead of crashing/failing.
- Unit tests cover filtering logic for blocklist and allowlist modes.
- Existing integrations remain green - default behavior is unchanged (no filtering) unless configured.
- Documentation is updated with examples.

### Area

Proxy / Routing

### Related Provider(s)

NVIDIA NIM, and any provider that does not support the full OpenAI-compatible parameter spec

### Additional Context

None

### Expected Test Plan

_No response_

### PR DESCRIPTION
**Pull request by ThongAccount**
## Summary

Add a config-driven **parameter denylist/allowlist** system that lets operators control which request parameters are stripped before being sent to upstream providers, plus **auto-learning** from upstream 400 errors.

When a provider like NVIDIA NIM rejects parameters (e.g. `thinking` → `400: Unsupported parameter(s): thinking`), the operator can now proactively block those params per provider or per model. The system also supports an allowlist that re-adds specific params after denylist stripping, and auto-learns new unsupported params from 400 errors.

**Capabilities added:**
- New DB module stores per-provider filter configs in `key_value` (namespace `provider_param_filters`), with in-memory cache + generation counter invalidation
- Extended `stripUnsupportedParams()` with `applyConfigFilters()` — applies provider-level denylist → model-level denylist → provider-level allowlist → model-level allowlist, all post-translation in the executor pipeline
- Auto-learn: when a 400 with `"Unsupported parameter: X"` is detected and either the global flag or per-provider `autoLearn` flag is enabled, the param is automatically persisted to the block list, stripped, and the request retried
- Global auto-learn flag acts as a master switch; per-provider flag works independently — either being on is sufficient
- Existing hardcoded `STRIP_RULES` unchanged — config rules compose on top
- API route `GET|PUT|DELETE /api/providers/[id]/param-filters` for programmatic access
- Provider-level `ProviderParamFilterSection` React component wired into the provider detail page with block/allow/autoLearn save/reset
- Model-level block/allow config in the existing ModelCompatPopover (compatibility popover on each model row)

## Related Issues

- Closes #6625

## Validation

- [ ] Run unit tests: `node --import tsx/esm --test tests/unit/param-filters-apply.test.ts tests/unit/param-filters-db.test.ts tests/unit/param-filters-auto-learn.test.ts`
- [ ] Run existing regression tests: `node --import tsx/esm --test tests/unit/provider-field-strips.test.ts tests/unit/groq-field-strip-wiring.test.ts`
- [ ] API CRUD: `curl -X PUT /api/providers/nvidia/param-filters` then `curl -X GET` to verify round-trip
- [ ] Global flag: `curl -X PUT /api/providers/__global__/param-filters -d '{"autoLearn":true}'` then verify via GET
- [ ] Reset: `curl -X DELETE /api/providers/nvidia/param-filters` and confirm empty config
- [ ] No config = no filtering (default behavior unchanged)

## Tests Added Or Updated

- `tests/unit/param-filters-apply.test.ts` — hardcoded rules regression + `applyConfigFilters()` direct tests (no DB dependency)
- `tests/unit/param-filters-db.test.ts` — CRUD on `key_value`, cache invalidation, full filter pipeline (DB-backed config → `stripUnsupportedParams()`), global auto-learn flag; 20 tests in isolated temp DB
- `tests/unit/param-filters-auto-learn.test.ts` — `UNSUPPORTED_PARAM_RE` regex matching + `detectUnsupportedParam()` edge cases

## Coverage Notes

Changes touch `open-sse/translator/paramSupport.ts`, `open-sse/config/providerFieldStrips.ts`, `open-sse/executors/base.ts`, `src/lib/db/paramFilters.ts`, `src/lib/localDb.ts`, and `src/lib/db/migrations/118_provider_param_filters.sql`. Covered by 28 new unit tests + existing `provider-field-strips.test.ts` and `groq-field-strip-wiring.test.ts` regression suites.

## Reviewer Notes

- **No feature flags** — opt-in by nature: no `key_value` row means no filtering, exactly like today
- **Allowlist semantics**: allowlist only restores params the client actually sent (from a pre-strip snapshot). It never introduces new params.
- **Auto-learn safety**: rate-limited (implicit via `strippedFields` set per request), logs at `info` level, requires either global flag or per-provider `autoLearn: true`
- **Global flag**: stored as `key = __global__` in the same `provider_param_filters` namespace; `isAutoLearnGloballyEnabled()` || `config.autoLearn` determines whether auto-learn fires
- **UI complete**: provider-level section on each provider detail page, model-level block/allow in the compatibility popover on each model row

### CODE REVIEW

File: open-sse/executors/base.ts [hunk 1]
```diff
@@ -1261,6 +1270,39 @@ export class BaseExecutor {
...
+              autoLearned &&
+              !strippedFields.has(autoLearned) &&
+              (transformedBody as Record<string, unknown>)[autoLearned] !== undefined
+            ) {
+              try {
+                const config = getParamFilterConfig(this.provider);
+                const shouldAutoLearn = isAutoLearnGloballyEnabled() || config?.autoLearn === true;
+                if (shouldAutoLearn) {
+                  strippedFields.add(autoLearned);
+                  addParamToBlocklist(this.provider, autoLearned);
```
[the remark this answers is not in the document]
Comment (ThongAccount) [reply open-sse/executors/base.ts]: Fixed in 0f319acb. Changed `addParamToBlocklist(this.provider, autoLearned)` to `addParamToBlocklist(this.provider, autoLearned, model)` so auto-learn scopes to the specific model that triggered the 400, preventing provider-level pollution.

File: open-sse/translator/paramSupport.ts [hunk 2]
```diff
@@ -56,15 +58,77 @@ export function stripUnsupportedParams<T>(
...
+  }
+
+  // 4. Model-level allowlist
+  if (modelCfg?.allow) {
+    for (const key of modelCfg.allow) {
+      if (key in snapshot) {
+        body[key] = snapshot[key];
+      }
+    }
+  }
```
[the remark this answers is not in the document]
Comment (ThongAccount) [reply open-sse/translator/paramSupport.ts]: Fixed in 0f319acb. Reordered applyConfigFilters so model-level operations run AFTER provider-level operations:
1. Provider denylist
2. Provider allowlist
3. Model denylist (overrides provider allowlist)
4. Model allowlist (most specific wins)

Added regression test verifying model denylist beats provider allowlist.

File: open-sse/executors/base.ts [hunk 3]
```diff
@@ -1261,6 +1270,39 @@ export class BaseExecutor {
...
+                  delete (transformedBody as Record<string, unknown>)[autoLearned];
+                  let retryBody = JSON.stringify(transformedBody);
+                  if (isClaudeCodeCompatible(this.provider) || this.provider === "claude") {
+                    retryBody = await signRequestBody(retryBody);
+                  }
+                  log?.info?.(
+                    "AUTO_LEARN",
+                    `Auto-learned "${autoLearned}" for provider ${this.provider} from 400 on ${url} — retrying`
+                  );
```
[the remark this answers is not in the document]
Comment (ThongAccount) [reply open-sse/executors/base.ts:1296]: Fixed in 0f319acb. Log message now includes model: `Auto-learned "X" for provider Y (model: Z) …`

**Comment by diegosouzapw** [comment]
Hi @ThongAccount, thanks for this — a config-driven param denylist/allowlist with auto-learn from upstream 400s is a genuinely useful generalization of the hardcoded STRIP_RULES we already have scattered across `providerFieldStrips.ts`/`paramSupport.ts`, and #6625 is a real pain point. The precedence logic (provider deny → provider allow → model deny → model allow) is well thought out and your 3 test files (33 tests) all pass green, with no regressions in the existing strip-rules suite.

We ran the full local gate set against the branch and found a handful of pre-merge issues we're going to fix directly in your branch (keeping your authorship) rather than bounce back and forth:

- `npm run typecheck:core` currently fails: `invalidateDbCache(NAMESPACE)` in `paramFilters.ts` passes a scope string that isn't part of `invalidateDbCache`'s accepted union — we'll drop that call (your local `filterCache = null` already handles cache invalidation for this module).
- The new `/api/providers/[id]/param-filters` route has no auth check at all — every sibling route under `providers/[id]/` gates on `requireManagementAuth()`, and there's no global middleware here to fall back on. We'll add it.
- The 3 handlers return raw `error.message` in the response body; our convention (see `docs/security/ERROR_SANITIZATION.md`) is to route every HTTP error response through `buildErrorBody()`/`sanitizeErrorMessage()` — we'll swap that in.
- The PUT body is hand-validated instead of going through a Zod schema like the sibling provider-update route — we'll add one.
- `ProviderDetailPageClient.tsx` and `src/lib/localDb.ts` both crossed our frozen file-size budget by a few lines — we'll extract/relocate to bring them back under.
- Three new i18n keys used in `ModelCompatPopover.tsx` (`compatParamFiltersLabel`, `compatBlockedParamsHint`, `compatAllowedParamsHint`) aren't registered in `en.json`, so they'll render as the raw key literal instead of your intended fallback text — we'll add them, and also wire `ProviderParamFilterSection.tsx`'s strings through `t()` for consistency with the rest of the dashboard.

None of this reflects on the core design — it's all mechanical cleanup. If you'd like to take a pass at any of these yourself before we do, just say so; otherwise we'll land the fixes and get this merged.

**Comment by ThongAccount** [comment]
@diegosouzapw, thanks for the review. I'll take a pass at the fixes myself:
- Removed the stray invalidateDbCache call
- Added requireManagementAuth() to the route
- Swapped error responses to buildErrorBody()
- Added Zod schema for PUT validation
- Registered the i18n keys in en.json and wired ProviderParamFilterSection through t()

The file-size budget items I'll leave to you since you have more context on the extraction patterns. Everything else and validation stats should be landing and updating shortly.

**Comment by ThongAccount** [comment]
All the fixes applied in aa1c9e3.

**Comment by diegosouzapw** [comment]
Merged — thank you, @ThongAccount! The config-driven param denylist/allowlist with auto-learn is a great generalization of our hardcoded STRIP_RULES, and your precedence design + 33 tests held up all the way through. We landed the review follow-ups directly in your branch (keeping your authorship): auth gate on the new route, error sanitization, Zod body validation, a typecheck fix, file-size/complexity gate compliance, and the missing i18n keys. Closes #6625 — ships in v3.8.47. 🚀
